src/dos/cdrom.cpp

```diff
@@ -18,6 +18,217 @@
 
 #include "cdrom.h"
 
+#include "../libs/sdlcd/SDL_cdrom.c"
+#if defined(WIN32)
+#define SDL_CDROM_WIN32
+#include "../libs/sdlcd/win32/SDL_syscdrom.c"
+#elif defined(LINUX)
+#define SDL_CDROM_LINUX
+#include "../libs/sdlcd/linux/SDL_syscdrom.c"
+#elif defined(MACOSX) && defined(__clang__)
+#define SDL_CDROM_MACOSX
+#include "../libs/sdlcd/macosx/SDL_syscdrom.c"
+#include "../libs/sdlcd/macosx/AudioFilePlayer.c"
+#include "../libs/sdlcd/macosx/AudioFileReaderThread.c"
+#include "../libs/sdlcd/macosx/CDPlayer.c"
+#include "../libs/sdlcd/macosx/SDLOSXCAGuard.c"
+#else
+#define SDL_CDROM_DUMMY
+#include "../libs/sdlcd/dummy/SDL_syscdrom.c"
+#endif
+
+// ******************************************************
+// SDL CDROM
+// ******************************************************
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "dosbox.h"
+#include "SDL.h"
+#include "support.h"
+#include "cdrom.h"
+
+CDROM_Interface_SDL::CDROM_Interface_SDL(void)
+{
+	driveID = 0;
+	oldLeadOut = 0;
+	cd = 0;
+}
+
+CDROM_Interface_SDL::~CDROM_Interface_SDL(void)
+{
+	StopAudio();
+	SDL_CDClose(cd);
+	cd = 0;
+}
+
+bool CDROM_Interface_SDL::SetDevice(char *path, int forceCD)
+{
+	char buffer[512];
+	strcpy(buffer, path);
+	upcase(buffer);
+
+	int num = SDL_CDNumDrives();
+	if ((forceCD >= 0) && (forceCD < num)) {
+		driveID = forceCD;
+		cd = SDL_CDOpen(driveID);
+		SDL_CDStatus(cd);
+		return true;
+	};
+
+	const char *cdname = 0;
+	for (int i = 0; i < num; i++) {
+		cdname = SDL_CDName(i);
+		if (strcmp(buffer, cdname) == 0) {
+			cd = SDL_CDOpen(i);
+			SDL_CDStatus(cd);
+			driveID = i;
+			return true;
+		};
+	};
+	return false;
+}
+
+bool CDROM_Interface_SDL::GetAudioTracks(uint8_t &stTrack, uint8_t &end, TMSF &leadOut)
+{
+	if (CD_INDRIVE(SDL_CDStatus(cd))) {
+		stTrack = 1;
+		end = cd->numtracks;
+		leadOut = frames_to_msf(cd->track[cd->numtracks].offset);
+	}
+	return CD_INDRIVE(SDL_CDStatus(cd));
+}
+
+bool CDROM_Interface_SDL::GetAudioTrackInfo(uint8_t track,
+                                            TMSF &start,
+                                            unsigned char &attr)
+{
+	if (CD_INDRIVE(SDL_CDStatus(cd))) {
+		start = frames_to_msf(cd->track[track - 1].offset);
+		attr = cd->track[track - 1].type << 4; // sdl uses 0 for audio
+		                                       // and 4 for data. instead
+		                                       // of 0x00 and 0x40
+	}
+	return CD_INDRIVE(SDL_CDStatus(cd));
+}
+
+bool CDROM_Interface_SDL::GetAudioSub(unsigned char &attr,
+                                      unsigned char &track,
+                                      unsigned char &index,
+                                      TMSF &relPos,
+                                      TMSF &absPos)
+{
+	if (CD_INDRIVE(SDL_CDStatus(cd))) {
+		track = cd->cur_track;
+		index = cd->cur_track;
+		attr = cd->track[track].type << 4;
```

Comment:
Is this a candidate for the new bit operations helpers?

Reply:
I guess a simple left shift like this may not really be included in the new bit operations helpers.